### PR TITLE
Implements estimate/compute_cost for AStar2D

### LINKED ISSUE #36990: AStar2D cannot override _estimate_cost or _compute_cost in GDScript
Godot version : 3.2.1, also present in 4.0

Issue description :
AStar provides a way to override its cost calculation from GDScript. Since AStar2D is a wrapper around it with an `AStar astar` member there is no way to override the `_evaluate_cost()` and `_compute_cost()` inside of it

Steps to reproduce :

`class CustomAStar2D :
	extends AStar2D
	
	func _estimate_cost(u, v):
		print("this isn't called")
		var dist = (get_point_position(u) - get_point_position(v)).abs()
		return dist.x + dist.y

	func _compute_cost(u, v):
                print("not called either")
		return _estimate_cost(u, v)`

I'm not sure if the best would be to rewrite AStar2D to use its own _solve() which might also slightly improve performance by removing useless calculations. Or completly rewrite AStar2D to not be a wrapper ? Or maybe there is a way to reach `astar._estimate_cost()` from its wrapper class ?

At least if it's intended that advance uses needing to overwrite those methods should use AStar regardless of 2D/3D, it should be documented here : https://docs.godotengine.org/en/3.2/classes/class_astar2d.html

### PR DESCRIPTION
**Pull request by Chaosus**
Fix #36990

### CODE REVIEW

File: core/math/a_star.h [hunk 1]
```diff
@@ -169,6 +170,11 @@ class AStar2D : public Reference {
 protected:
 	static void _bind_methods();
 
+	bool _solve(AStar::Point *begin_point, AStar::Point *end_point);
+
+	virtual float _estimate_cost(int p_from_id, int p_to_id);
+	virtual float _compute_cost(int p_from_id, int p_to_id);
```
Comment (aaronfranke) [inline core/math/a_star.h]: These should be `real_t` since the values come from `distance_to` which uses `real_t`.

**Review by aaronfranke — DISMISSED** [review]

File: core/math/a_star.h (hunk 1 above)
Comment (Chaosus) [reply core/math/a_star.h]: Fixed...

**Comment by akien-mga** [comment]
Thanks!

**Comment by eltharionx** [comment]
Not sure where else to ask this, but does this fix plan on being released on the next Godot 3.2.x release?

**Comment by Calinou** [comment]
@eltharionx This pull request has the `cherrypick:3.2` label, which means it will be cherry-picked for Godot 3.2.3 when Akien has time.

**Comment by akien-mga** [comment]
Cherry-picked for 3.2.2.

**Pull request merged**